Client/Frontend/Browser/SearchEngines.swift

```diff
@@ -140,27 +140,55 @@ class SearchEngines {
         }
     }
 
+    /// Return all possible paths for a language identifier in the order of most specific to least specific.
+    /// For example, zh-Hans-CN with a default of en will return [zh-Hans-CN, zh-CN, zh, en]. The fallback
+    /// identifier must be a known one that is guaranteed to exist in the SearchPlugins directory.
+    class func directoriesForLanguageIdentifier(languageIdentifier: String, basePath: NSString, fallbackIdentifier: String) -> [String] {
+        var directories = [String]()
+        let components = languageIdentifier.componentsSeparatedByString("-")
+        if components.count == 1 {
+            // zh
+            directories.append(languageIdentifier)
+        } else if components.count == 2 {
+            // zh-CN
+            directories.append(languageIdentifier)
+            directories.append(components[0])
+        } else if components.count == 3 {
+            directories.append(languageIdentifier)
+            directories.append(components[0] + "-" + components[2])
+            directories.append(components[0])
+        }
+        if !directories.contains(fallbackIdentifier) {
+            directories.append(fallbackIdentifier)
+        }
+        
+        return directories.map { (path) -> String in
+            return basePath.stringByAppendingPathComponent(path)
+        }
+    }
+
     // Get all known search engines, with the default search engine first, but the others in no
     // particular order.
     class func getUnorderedEngines() -> [OpenSearchEngine] {
         let pluginBasePath: NSString = (NSBundle.mainBundle().resourcePath! as NSString).stringByAppendingPathComponent("SearchPlugins")
-        let language = NSLocale.preferredLanguages().first!
+        let languageIdentifier = NSLocale.preferredLanguages().first!
         let fallbackDirectory: NSString = pluginBasePath.stringByAppendingPathComponent("en")
 
-        // Look for search plugins in the following order:
-        //   1) the full language ID
-        //   2) the language ID without the dialect
-        //   3) the fallback language (English)
-        // For example, "fr-CA" would look for plugins in this order: 1) fr-CA, 2) fr, 3) en.
-        var searchDirectory = pluginBasePath.stringByAppendingPathComponent(language)
-        if !NSFileManager.defaultManager().fileExistsAtPath(searchDirectory) {
-            let languageWithoutDialect = language.componentsSeparatedByString("-").first!
-            searchDirectory = pluginBasePath.stringByAppendingPathComponent(languageWithoutDialect)
-            if language == languageWithoutDialect || !NSFileManager.defaultManager().fileExistsAtPath(searchDirectory) {
-                searchDirectory = fallbackDirectory as String
+        print(directoriesForLanguageIdentifier(languageIdentifier, basePath: pluginBasePath, fallbackIdentifier: "en"))
```

Comment:
Nit: should probably remove this to reduce noise, or at least put it behind a `log.debug()`.